### PR TITLE
Fix wrong message error when integer > max_int

### LINKED ISSUE #160: [PE-D] Entering huge numbers (e.g 2^31) causes wrong error message to appear
Entering a huge number (2^31) causes the "Carbohydrate values should only contain positive numbers" error message to appear instead of the "Carbohydrate amount can only take in value larger than 0 and less than 1000" error message.

-------------
Labels: `severity.VeryLow` `type.FunctionalityBug`
original: tjwjoe/ped#8

### LINKED ISSUE #161: [PE-D] Inaccurate error message
10. Error message "<field> should only contain positive numbers" would be more accurate if it said "non-negative integers", because it accepts 0 and doesn't accept floating point values.

-------------
Labels: `severity.VeryLow` `type.FunctionalityBug`
original: tjwjoe/ped#9

### LINKED ISSUE #179: [PE-D] Error message for invalid integer inputs does not match the error
With large integer input, it has probably exceeded the integer range, however, it is still considered as a valid positive integer, but the error message does not clearly explain why I got this error

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: minzzelo/ped#5

### PR DESCRIPTION
**Pull request by aidoxe-123**
close #179 , #161 , #160

### CODE REVIEW

File: src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -31,6 +31,8 @@
 public class ParserUtil {
 
     public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
+    public static final String MESSAGE_INDEX_TOO_LARGE = "Index must be smaller than 2^31.";
```
Comment (JoeyChenSmart) [inline src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: I think our limit for values is 1000, there's a function in `MacroNutrient` that says
```java
private String getMessageConstraint() {
        return this.getMacronutrientType() + " amount can only take in value larger than 0 and less than 1000";
    }
```

Also I think we should make all the error messages use this function somehow, so we have DRY. see #215

Comment (aidoxe-123) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: I think this is for the index, not the macronutrient value?

Comment (JoeyChenSmart) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: I'm quite sure its for all of those though? either way we have to limit all of them to much lower than MAX_INT

Comment (chewypiano) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: > I think this is for the index, not the macronutrient value?

As Joey mentioned correctly, we are limiting macronutrient values to 1000g each. Firstly because exceeding 1kg of any type of nutrient for a single dish does not make much sense so limiting it to 1000g does not limit its utility. Second and more importantly, exceeding 1000g of macronutrient will increase total calorie count by a lot, which in turn will cause the total calorie count to reach integer overflow easily. 

Basically, yeah we should limit protein, carb, fat value to 1000g.

Comment (aidoxe-123) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: ok, I will limit protein, carb, and fat value to 1000. But do we really need to also limit index value to 1000 too?

File: docs/UserGuide.md [hunk 2]
```diff
@@ -164,9 +164,8 @@ Examples:
 
 :information_source:
  * The default value for `PROTEIN`, `FATS` and `CARBS` is 0. The default date is the day in which the food item is added.<br>
- * Values of `PROTEIN`, `FATS` and `CARBS` must be a non-negative integer.<br>
+ * Values of `PROTEIN`, `FATS` and `CARBS` must be a non-negative integer that is smaller than 2^31.<br>
```
Comment (Jh123x) [inline docs/UserGuide.md]: It must be smaller than 1000 rather than 2^31. Same for the ones below

File: docs/UserGuide.md [hunk 3]
```diff
@@ -184,7 +183,7 @@ Format: `tag INDEX -t TAG_NAME`
 * Tags `TAG_NAME` for food item at the specified `INDEX`.
 * `TAG_NAME` is case-sensitive.
 * The index refers to the index number shown in the displayed food list.
-* The index **must be a positive integer** 1, 2, 3, …​
+* The index **must be a positive integer** 1, 2, 3, …​ that is smaller than 2^31
```
Comment (Jh123x) [inline docs/UserGuide.md:195]: It should be no larger than the size of the list instead of smaller than 2^31

**Review by Jh123x — CHANGES_REQUESTED** [review]
Mostly LGTM
Need some minor changes in UG.
Your implementation is q clean too without many repeats but it involves the creation of a Macronutrient when the value is needed.

It might be better to have a static reference to a default macronutrient to be accessed instead of creating a new one to be accessed everytime

File: src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java (hunk 1 above)
Comment (Jh123x) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: I don't think its for index. It only applies to MacroNutrients

File: docs/UserGuide.md (hunk 3 above)
Comment (aidoxe-123) [reply docs/UserGuide.md:195]: For this, assume that the user inputs some number that is larger MAX_INT. What message should we return in that case: the index is larger than 2^31, or the index is larger than the size of the list?

File: docs/UserGuide.md (hunk 2 above)
Comment (aidoxe-123) [reply docs/UserGuide.md:167]: > Your implementation is q clean too without many repeats but it involves the creation of a Macronutrient when the value is needed.
> 
> It might be better to have a static reference to a default macronutrient to be accessed instead of creating a new one to be accessed everytime

You got a point here. I've changed it back to the original one

**Comment by aidoxe-123** [comment]
> Your implementation is q clean too without many repeats but it involves the creation of a Macronutrient when the value is needed.
> 
> It might be better to have a static reference to a default macronutrient to be accessed instead of creating a new one to be accessed everytime

You got a point here. I've changed it back to the original one

File: src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -30,7 +30,7 @@
  * Contains utility methods used for parsing strings in the various *Parser classes.
  */
 public class ParserUtil {
-
+    private static final String VALIDATION_REGEX = "[0-9]+";
```
Comment (Jh123x) [inline src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: As this is only change used for the index, is it better to store this within index instead of ParserUtil?

**Review by Jh123x — APPROVED** [review]
LGTM

File: docs/UserGuide.md (hunk 3 above)
Comment (Jh123x) [reply docs/UserGuide.md:195]: hmm, then it should be the minimum of the 2?
we can also change it to long just in case too

File: src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java (hunk 4 above)
Comment (aidoxe-123) [reply src/main/java/jimmy/mcgymmy/logic/parser/ParserUtil.java]: ok, I will move it into parseIndex
